rs/consensus/vetkd/src/lib.rs

```diff
@@ -0,0 +1,390 @@
+            return vec![];
+        };
+
+        let delivered_ids = parse_past_payload_ids(past_payloads, &self.log);
+        let payload = self.get_vetkd_payload_impl(
+            context.time,
+            config,
+            state.get_ref(),
+            delivered_ids,
+            max_size,
+        );
+        vetkd_payload_to_bytes(payload, max_size)
+    }
+
+    fn validate_payload(
+        &self,
+        height: Height,
+        context: &ProposalContext,
+        payload: &[u8],
+        past_payloads: &[PastPayload],
+    ) -> Result<(), PayloadValidationError> {
+        // Empty payloads are always valid
+        if payload.is_empty() {
+            return Ok(());
+        }
+
+        let config = self.get_chain_key_config_if_enabled(height)?;
+
+        let state = match self
+            .state_reader
+            .get_state_at(context.validation_context.certified_height)
+        {
+            Ok(state) => state,
+            Err(err) => {
+                return validation_failed_err(VetKdPayloadValidationFailure::StateUnavailable(err));
+            }
+        };
+
+        let delivered_ids = parse_past_payload_ids(past_payloads, &self.log);
+        let payload = bytes_to_vetkd_payload(payload)
+            .map_err(|e| invalid_artifact(InvalidVetKdPayloadReason::DeserializationFailed(e)))?;
+
+        self.validate_vetkd_payload_impl(
+            payload,
+            context.validation_context.time,
+            config,
+            state.get_ref(),
+            delivered_ids,
+        )
+    }
+}
+
+impl IntoMessages<Vec<ConsensusResponse>> for VetKdPayloadBuilderImpl {
+    fn into_messages(payload: &[u8]) -> Vec<ConsensusResponse> {
+        let messages = bytes_to_vetkd_payload(payload)
+            .expect("Failed to parse a payload that was already validated");
+
+        messages
+            .vetkd_agreements
+            .into_iter()
+            .map(|(id, response)| {
+                ConsensusResponse::new(
+                    id,
+                    match response {
+                        VetKdAgreement::Success(data) => ResponsePayload::Data(data),
+                        VetKdAgreement::Reject(error_code) => {
+                            ResponsePayload::Reject(match error_code {
+                                VetKdErrorCode::TimedOut => RejectContext::new(
+                                    RejectCode::CanisterError,
+                                    "VetKD request expired",
+                                ),
+                                VetKdErrorCode::InvalidKey => RejectContext::new(
+                                    RejectCode::CanisterError,
+                                    "Invalid key_id in VetKD request",
+                                ),
+                            })
+                        }
+                    },
+                )
+            })
+            .collect()
+    }
+}
+
+fn maybe_reject(
```

Comment:
The name is a bit too general.
Maybe `reject_if_invalid` or similar.
Also a description would help